src/vs/workbench/contrib/positronPlots/test/browser/positronPlotsService.test.ts

```diff
@@ -0,0 +1,230 @@
+import { IViewsService } from 'vs/workbench/services/views/common/viewsService';
+import { TestViewsService, workbenchInstantiationService } from 'vs/workbench/test/browser/workbenchTestServices';
+
+suite('Positron - Plots Service', () => {
+
+	const disposables = ensureNoDisposablesAreLeakedInTestSuite();
+	let plotsService: PositronPlotsService;
+	let runtimeSessionService: TestRuntimeSessionService;
+
+	setup(() => {
+		const instantiationService = workbenchInstantiationService(undefined, disposables);
+		runtimeSessionService = disposables.add(instantiationService.createInstance(TestRuntimeSessionService));
+		instantiationService.stub(IRuntimeSessionService, runtimeSessionService);
+		instantiationService.stub(IPositronWebviewPreloadService, disposables.add(instantiationService.createInstance(PositronWebviewPreloadService)));
+		instantiationService.stub(IPositronIPyWidgetsService, disposables.add(instantiationService.createInstance(PositronIPyWidgetsService)));
+		instantiationService.stub(IViewsService, new TestViewsService());
+
+		plotsService = disposables.add(instantiationService.createInstance(PositronPlotsService));
+	});
+
+	async function createSession() {
+		const session = disposables.add(new TestLanguageRuntimeSession(LanguageRuntimeSessionMode.Console));
+		runtimeSessionService.startSession(session);
+
+		await timeout(0);
+
+		const out: {
+			session: TestLanguageRuntimeSession;
+			plotClient: IPositronPlotClient | undefined;
+		} = {
+			session, plotClient: undefined,
+		};
+
+		disposables.add(session.onDidCreateClientInstance(client => out.plotClient = {
+			id: client.client.getClientId(),
+			metadata: {} as IPositronPlotMetadata,
+		} as IPositronPlotClient));
+
+		return out;
+	}
+
+	test('history policy: change history policy', () => {
+		plotsService.selectHistoryPolicy(HistoryPolicy.AlwaysVisible);
+		assert.strictEqual(plotsService.historyPolicy, HistoryPolicy.AlwaysVisible);
+
+		plotsService.selectHistoryPolicy(HistoryPolicy.Automatic);
+		assert.strictEqual(plotsService.historyPolicy, HistoryPolicy.Automatic);
+
+		plotsService.selectHistoryPolicy(HistoryPolicy.NeverVisible);
+		assert.strictEqual(plotsService.historyPolicy, HistoryPolicy.NeverVisible);
+	});
+
+	test('history policy: change event', async () => {
+		let historyPolicyChanged = 0;
+
+		const didChangeHisotryPolicy = new Promise<void>((resolve) => {
+			const disposable = plotsService.onDidChangeHistoryPolicy((e) => {
+				historyPolicyChanged++;
+				resolve();
+			});
+			disposables.add(disposable);
+		});
+
+		// no event since 'Automatic' is the default
+		plotsService.selectHistoryPolicy(HistoryPolicy.Automatic);
+
+		// event occurs when changing to 'AlwaysVisible'
+		plotsService.selectHistoryPolicy(HistoryPolicy.AlwaysVisible);
+
+		await raceTimeout(didChangeHisotryPolicy, 100, () => assert.fail('onDidChangeHistoryPolicy event did not fire'));
```

Comment:
Does this actually fail the test given that `assert.fail` is called in a callback?

Reply:
Yes, this does work but only if the callback doesn't execute within 100ms. This is used upstream and I think it is okay to use for tests.